libsolidity/codegen/ArrayUtils.cpp

```diff
@@ -1055,28 +1055,27 @@ void ArrayUtils::accessIndex(ArrayType const& _arrayType, bool _doBoundsCheck, b
 	switch (location)
 	{
 	case DataLocation::Memory:
+		// stack: <base_ref> <index>
+		if (!_arrayType.isByteArray())
+			m_context << u256(_arrayType.memoryHeadSize()) << Instruction::MUL;
```

Comment:
Is this correct for arrays that are dynamically encoded? E.g.
```
struct S { uint256 a; uint256 b; }
function f(S[] calldata s) {...}
```
It might be that ``memoryHeadSize()`` will be ``64`` in this case, but we want to multiply by ``32`` instead - I'm not sure, but we should check.

Reply:
Similarly ``function f(uint[2][] calldata arr) {...}``.

Reply:
*If* this is incorrect, it'd be spooky that we don't have test failures, but we might just not have enough tests for these cases yet...

Reply:
I'm probably mistaken here and this is fine, but I'll need to recheck.

Reply:
Hmm... good catch, I guess it should be `_arrayType.baseType()->memoryHeadSize()` instead of `_arrayType.memoryHeadSize()`, right? 

> It might be that memoryHeadSize() will be 64 in this case

Because it is a reference type, it should be 32
https://github.com/ethereum/solidity/blob/dd04a35c0e2d2111ff165013b0c7fe1a31beb3a7/libsolidity/ast/Types.h#L636

> If this is incorrect, it'd be spooky that we don't have test failures

I think the reason is that currently in solidity codebase, all invocation to `memoryHeadSize()` would returns 32. For types occupy less than 32 bytes we pads, and types occupy more than 32 bytes are all reference type. This is just my guess though, we should check that this is indeed the case.

Reply:
Ah right - I overlooked the override of ``memoryHeadSize`` in ``ReferenceType`` and only looked at the implementation in ``Type`` that falls back to ``calldataEncodedSize``. You're right: since ``memoryHeadSize()`` always returns ``32`` here, it should be fine in all cases.

Reply:
Or should it :-)? I'll take some more time looking into it :-).

Reply:
It still looks somewhat wrong to me and feels like ``_arrayType.baseType()->memoryHeadSize()`` should be correct, if the base type is statically encoded and ``32`` should be correct, if the base type is dynamically encoded - so exactly what the ``CallData`` case does below... but I'm probably missing something myself right now and so far I haven't been able to actually break this...

Reply:
This shouldn't be an issue for this PR in any case, since it did the same thing before and I'm probably just not seeing why this is correct right now...

Reply:
Ah ok, now I see - I was thinking about calldata layout the whole time, for which statically sized arrays are encoded in-place, whereas memory arrays always use tail pointers anyways, so this *is* correct - nevermind :-)!